core/module_graph.cc

```diff
@@ -81,19 +84,133 @@ void ModuleGraph::UpdateSingleTaskGraph(Module *task_module) {
   }
 }
 
+void ModuleGraph::PropagateIGatePriority(
+    bess::IGate *igate, std::unordered_set<bess::IGate *> &visited_igates,
+    uint32_t priority) {
+  if (igate->module()->is_task()) {
+    return;
+  } else {
+    std::vector<bess::OGate *> ogates = igate->module()->ogates();
+    for (size_t i = 0; i < ogates.size(); i++) {
+      if (!ogates[i]) {
+        break;
+      }
+
+      bess::IGate *next_igate = ogates[i]->igate();
+      if (visited_igates.count(next_igate) != 0 ||  // This is a loop or
+          next_igate->priority() >= priority) {     // visited by longer path
+        continue;
+      }
+
+      visited_igates.insert(next_igate);
+      next_igate->SetPriority(priority);
+      priority++;
```

Comment:
Is there a reason to increment `priority` and then decrement again?  Why not replace these three lines with:

`PropagateIGatePriority(next_igate, visited_igates, priority+1);`